aeppl/mixture.py

```diff
@@ -300,6 +302,67 @@ def mixture_replace(fgraph, node):
     return [new_mixture_rv]
 
 
+@local_optimizer((Elemwise,))
+def switch_mixture_replace(fgraph, node):
+    r"""Identify mixture sub-graphs defined via a Switch Op and replace them with
+    a place-holder MixtureRV Op
+    """
+    try:
+        if not isinstance(node.op.scalar_op, Switch):
+            return None  # pragma: no cover
+    except AttributeError:
+        return None
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return None  # pragma: no cover
+
+    old_mixture_rv = node.default_output()
+
+    if len(node.inputs) != 3:
+        return None  # graph structure should be [index, component1, component2]
+
+    mixing_indices = node.inputs[0]
+    mixture_res = node.inputs[1:]
+
+    # We loop through mixture components and collect all the array elements
+    # that belong to each one (by way of their indices).
+    mixture_rvs = []
+    for i, component_rv in enumerate(mixture_res):
+
+        # We create custom types for the mixture components and assign them
+        # null get_measurable_outputs dispatches so that they aren't
+        # erroneously encountered in places like factorized_joint_logprob.
+        new_node = assign_custom_measurable_outputs(component_rv.owner)
+        out_idx = component_rv.owner.outputs.index(component_rv)
+        new_comp_rv = new_node.outputs[out_idx]
+        mixture_rvs.append(new_comp_rv)
+
+    # Replace this sub-graph with a MixtureRV
+    mix_op = MixtureRV(
+        2,
+        old_mixture_rv.dtype,
+        old_mixture_rv.broadcastable,
+    )
+    new_node = mix_op.make_node(*([NoneConst, mixing_indices] + mixture_rvs))
+
+    new_mixture_rv = new_node.default_output()
+
+    if aesara.config.compute_test_value != "off":
+        # We can't use MixtureRV to compute a test value; instead, we'll use
+        # the original node's test value.
+        if not hasattr(old_mixture_rv.tag, "test_value"):
+            compute_test_value(node)
+
+        new_mixture_rv.tag.test_value = old_mixture_rv.tag.test_value
+
+    if old_mixture_rv.name:
+        new_mixture_rv.name = f"{old_mixture_rv.name}-mixture"
+
+    return [new_mixture_rv]
```

Comment:
Thinking sort of out loud here. I have made progress on how to rewrite nested switches into a stack and I'm at the point where I should test my logic on more graphs (starting in NumPy).

However, I believe that I should make use of `mixture_replace` here. Should I be returning `mixture_replace(fgraph, node)` with a newly constructed `fgraph`? If so, I believe that the pseudo-code would look something like this:

```python
fgraph = FunctionGraph(...)
new_node = at.stack([component1, component2, ...])

return mixture_replace(fgraph, new_node)
```

Does this approach seem reasonable?

Reply:
Your rewrite should not have to construct a new fgraph (it takes it as input). I guess you can call `mixture_replace` directly, but if you return just the `new_node`, the `mixture_replace` will later be called on that node automatically anyway. @brandonwillard any preference here for directly calling `mixture_replace`?

Reply:
If we believe there will be no other rewrite that could/should be applied to the intermediate results, then, sure, calling `mixture_replace` directly is fine.  It might mess with some profiling results, but nothing serious.